website/blog/2024-08-29-autogen-pyfunc/index.md

```diff
@@ -0,0 +1,596 @@
+
+    return img_data
+
+def extract_img(agent: Agent) -> PIL.Image:
+    """
+    Extracts an image from the last message of an agent and converts it to a PIL image.
+
+    This function searches the last message sent by the given agent for an image tag,
+    extracts the image data, and then converts this data into a PIL (Python Imaging Library) image object.
+
+    Parameters:
+        agent (Agent): An instance of an agent from which the last message will be retrieved.
+
+    Returns:
+        PIL.Image: A PIL image object created from the extracted image data.
+
+    Note:
+    - The function assumes that the last message contains an <img> tag with image data.
+    - The image data is extracted using a regular expression that searches for <img> tags.
+    - It's important that the agent's last message contains properly formatted image data for successful extraction.
+    - The `_to_pil` function is used to convert the extracted image data into a PIL image.
+    - If no <img> tag is found, or if the image data is not correctly formatted, the function may raise an error.
+    """
+    last_message = agent.last_message()["content"]
+
+    if isinstance(last_message, str):
+        img_data = re.findall("<img (.*)>", last_message)[0]
+    elif isinstance(last_message, list):
+        # The GPT-4V format, where the content is an array of data
+        assert isinstance(last_message[0], dict)
+        img_data = last_message[0]["image_url"]["url"]
+
+    pil_img = get_pil_image(img_data)
+    return pil_img
+
+class DALLEAgent(ConversableAgent):
+    def __init__(self, name, llm_config: dict, **kwargs):
+        super().__init__(name, llm_config=llm_config, **kwargs)
+
+        api_key = os.getenv("OPENAI_API_KEY")
+        self._dalle_client = OpenAI(api_key=api_key)
+        self.register_reply([Agent, None], DALLEAgent.generate_dalle_reply)
+
+    def send(
+        self,
+        message: Union[Dict, str],
+        recipient: Agent,
+        request_reply: Optional[bool] = None,
+        silent: Optional[bool] = False,
+    ):
+        # override and always "silent" the send out message;
+        # otherwise, the print log would be super long!
+        super().send(message, recipient, request_reply, silent=True)
+
+    def generate_dalle_reply(self, messages: Optional[List[Dict]], sender: "Agent", config):
+        """Generate a reply using OpenAI DALLE call."""
+        client = self._dalle_client if config is None else config
+        if client is None:
+            return False, None
+        if messages is None:
+            messages = self._oai_messages[sender]
+
+        prompt = messages[-1]["content"]
+        # TODO: integrate with autogen.oai. For instance, with caching for the API call
+        img_data = dalle_call(
+            client=client,
+            model="dall-e-3",
+            prompt=prompt,
+            size="1024x1024",  # TODO: the size should be flexible, deciding landscape, square, or portrait mode.
```

Comment:
let's remove this. The values that can be specified are an enumeration. No need to complicate the example with that. hard-coding the spec is fine.